rust/kernel/fs.rs

```diff
@@ -44,6 +93,54 @@ impl<T: Type + ?Sized> Tables<T> {
         }
     }
 
+    unsafe extern "C" fn parse_param_callback(
+        fc: *mut bindings::fs_context,
+        param: *mut bindings::fs_parameter,
+    ) -> core::ffi::c_int {
+        from_kernel_result! {
+            // SAFETY: The callback contract guarantees that `fc` is valid.
+            let ptr = unsafe { (*fc).fs_private };
+
+            // SAFETY: The value of `ptr` (coming from `fs_private` was initialised in
+            // `init_fs_context_callback` to the result of an `into_pointer` call. Since the
+            // context is valid, `from_pointer` wasn't called yet, so `ptr` is valid. Additionally,
+            // the callback contract guarantees that callbacks are serialised, so it is ok to
+            // mutably reference it.
+            let mut data = unsafe { borrow_mut::<<T::Context as Context<T>>::Data>(ptr) };
```

Comment:
To fulfill the safety guarantee, we also need to make sure `&data` won't be `Copy` to another thread, right? Because `data` is `Sync`, allowing copy `&data` breaks the safety guarantee:

> no other concurrent users of the pointer run at least until the returned `ScopeGuard` is dropped.

Reply:
I think it's better if we have a `PointerWrapper::borrow_mut`

Reply:
> To fulfill the safety guarantee, we also need to make sure `&data` won't be `Copy` to another thread, right? Because `data` is `Sync`, allowing copy `&data` breaks the safety guarantee:
> 
> > no other concurrent users of the pointer run at least until the returned `ScopeGuard` is dropped.

It's ok to have `&data` in other threads as long as it is derived from this `data` (e.g., we call a function that takes a `&data` as argument and sends it out to other threads but guarantees that they are done before returning). What needs to be avoided is `&data` or `&mut data` acquired through other means.

> I think it's better if we have a `PointerWrapper::borrow_mut`

I wanted to avoid having each implementation replicate the same functionality, but thinking about it, we can just provide a default implementation and have everyone use it. Will do.

Reply:
(BTW, I expanded the comment to include the fact that we can have `&data` derived from `data`.)